VideoLocker/src/main/java/org/edx/mobile/view/CourseDetailFragment.java

```diff
@@ -260,4 +282,120 @@ private ViewHolder createCourseDetailFieldViewHolder(LayoutInflater inflater, Li
         TextView rowFieldName;
         TextView rowFieldText;
     }
+
+
+    /**
+     * Sets the onClickListener and the text for the enrollment button.
+     *
+     * If the current course is found in the list of cached course enrollment list, the button will
+     * be for viewing a course, otherwise, it will be used to enroll in a course. One clicked, user
+     * is then taken to the dashboard for target course.
+     */
+    private void configureEnrollButton() {
+        ServiceManager api = environment.getServiceManager();
+        // This call should already be cached, if not, set button as if not enrolled.
+        try {
+            List<EnrolledCoursesResponse> enrolledCoursesResponse = api.getEnrolledCourses(true);
+            for (EnrolledCoursesResponse course : enrolledCoursesResponse) {
+                if (course.getCourse().getId().equals(courseDetail.course_id)) {
+                    mEnrolled = true;
+                }
+            }
+        } catch (Exception ex) {
+            logger.debug("Unable to get cached enrollments list");
+        }
+
+        if (mEnrolled) {
+            mEnrollButton.setText(R.string.view_course_button_text);
+        } else {
+            mEnrollButton.setText(R.string.enroll_now_button_text);
+        }
+
+        mEnrollButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                if (!mEnrolled) {
+                    enrollInCourse();
+                } else {
+                    openCourseDashboard();
+                }
+            }
+        });
+    }
+
+    /**
+     * Enroll in a course, then send a local broadcast to notify
+     * {@link org.edx.mobile.view.MyCoursesListActivity} to refresh when it is resumed. Then open
+     * the course dashboard of the enrolled course.
+     */
+    public void enrollInCourse() {
+        EnrollForCourseTask enrollForCourseTask = new EnrollForCourseTask(getActivity(),
+                courseDetail.course_id, emailOptIn) {
+            @Override
+            public void onSuccess(Boolean result) {
+                if (result != null && result) {
+                    mEnrolled = true;
+                    logger.debug("Enrollment successful:" + courseDetail.course_id);
+
+                    mEnrollButton.setText(R.string.view_course_button_text);
+                    Toast.makeText(getContext(), R.string.you_are_now_enrolled, Toast.LENGTH_SHORT).show();
+
+                    signalMyCourseListToRefresh();
```

Comment:
I would swap the order here; Since the UI needs the `GetEnrolledCourseTask` call to finish, it would be better to initiate that request first, so we can show the course dashboard ASAP.